v3

brendt opened the pull request:
This is a complete overhaul and simplification of the package.

brendt commented:
Some questions:

- Do we want to support `$enum->isDraft()`? There's no autocompletion on those functions, and it's the same as `$enum->equals(MyEnum::DRAFT())`
- Do we want support for enum-specific methods. I'd say no: if you need specific functionality, you need to use a state
- What about `equals` and `isAny`? They could be merged in the same method: `function equals(Enum ...$others)`. If we want to keep these separated, maybe `equalsAny` is a better name?

Gummibeer commented:
Hey,

I will respond to the three points - but am also looking forward to our call next week. 😊

* I know that this overloaded method adds some complexity because of the PHP strange `__call()` and `__callStatic()` behavior. But we use them some times and like in the docs they can be completed via `@method` php-doc. So I won't fight with you to keep them, but it's one of the features I/we use - so would be cool to keep.
* I'm fine with dropping enum custom methods support. I don't see any need for them and for me it's also the wrong place for business logic. But at the end only public static methods produce problems.
* For me these methods are needed - we can talk about naming and unify them or switch to more common naming. But the equality method is needed because PHP hasn't any way to compare two objects by using defined properties. And the any is good for checks like `isWeekday()` or `isWeekend()` on a weekday enum. But we could also put them into one - even if I think that this will only increase complexity of this method and I don't see any benefit of doing it in one method?

For me the mapping via constants is the most used way to map values. But all the static method stuff isn't used by me.

AlexVanderbist commented:
Same 3 points:

* Personally I've always used the `->equals()` method to do this, that way it's the same across the entire codebase. However, `isDraft()` seems like a quick and nice feature to have without adding too much complexity or maintenance.
* I agree with Tom here; it's not the place for business logic imo. I used them once for custom mappings but in hindsight I would've preferred to do this mapping in a different place.
* I don't mind the idea of merging the two as `equals()`. How would it increase complexity?

Gummibeer commented:
> However, isDraft() seems like a quick and nice feature to have without adding too much complexity or maintenance.

It's only ~5 lines of code in two methods. So it's not this much effort, and for hardcoded comparisons I prefer this method naming - like `$status->isDraft()`

> I used them once for custom mappings but in hindsight I would've preferred to do this mapping in a different place.

Can agree here, I've also misused enums once with a lot of custom methods and logic for an unit conversion system. After I forgot which enum does what I've refactored it into custom classes and now it's clean and works.
So I'm totally fine with dropping support for custom methods as soon as they conflict with any logic and would require a lot of work to fix. But on the other side we already have all the logic now and the only thing we do not support are custom `public static` methods. So I wouldn't want to drop the simple public method support - primary because it's only one constant that prevents using them as enum values.
So I wouldn't like to change the current logic regarding custom methods? Except we drop public static methods as enum values at all? I never used them and for me the doc-tags and constant mappings are enough.

> I don't mind the idea of merging the two as equals(). How would it increase complexity?

You are right, by using `...$enums` we always get an array we can loop through and return if one matches.
Because there's no way for a `equalsAll(Enumerable[])` method I'm fine with merging them together and do a `equalsAny()` check.

@brendt I've seen that you've dropped Index support and renamed `name` to `value` and `value` to `label`.
In general it's okay for to rename the two string representations - but why did you've dropped index support? The index/order/sort/int of an enum is one of the fundamental parts for an enum. 🤔 And for more performant database queries the integer is also good.

brendt commented:
- Add value map support
- Add `::make()` method
- Separate PRs for easier review

Gummibeer commented:
As a conclusion this will be a super breaking release. 😄 
A lot of method drops, renaming and drop support for several ways of definition.

Was a great call/discussion @brendt - looking forward to the separated PRs!

brendt commented:
It's difficult to split this in separate PRs, since only changing the resolve methods kind of means you need to change everything. Instead I'm going to explain the changes in this PR with some code samples, because I agree that the diff is too large to comprehend. I'm going to make some more changes first, and afterwards I'll add some comments.

brendt commented:
So let's go through the changes in depth. 

The biggest change is of course to the `resolveDefinition` logic. v3 will only support enums that are listed in the doc blocks, this means that value-specific anonymous classes don't exist anymore. If people want this feature, they should look at using [states](https://github.com/spatie/laravel-model-states) instead.

This change significantly decreased the complexity of resolving enum definitions:

```php
private static function resolveDefinition(): array
    {
        $className = static::class;

        if (static::$definitionCache[$className] ?? null) {
            return static::$definitionCache[$className];
        }

        $reflectionClass = new ReflectionClass($className);

        $docComment = $reflectionClass->getDocComment();

        preg_match_all('/@method static self ([\w_]+)\(\)/', $docComment, $matches);

        $definition = [];

        $valueMap = static::values();

        $labelMap = static::labels();

        foreach ($matches[1] as $methodName) {
            $definition[$methodName] = new EnumDefinition(
                $methodName,
                $valueMap[$methodName] ?? $methodName,
                $labelMap[$methodName] ?? $methodName,
            );
        }

        return static::$definitionCache[$className] ??= $definition;
    }
```

This is what happens when resolving the enum's definition:

- Check whether the definition is cached, if so, return that
- Parse the class doc blocks looking for `@method static self` methods
- Add each method as a definition entry

I've added the concept of `EnumDefinition`, a value object that contains three things:

- The method name that constructs that enum
- The actual value
- The actual label

By default, the value and label of an enum are its method name, but they can be overridden. These concepts were previously known as `const MAP_INDEX` and `const MAP_VALUE`, but are now methods that can be overridden in the concrete enum classes. They are called `values` and `label`, and look like this:

```php
/**
 * @method static self A()
 * @method static self B()
 */
class EnumAsArray extends Enum
{
    protected static function labels(): array
    {
        return [
            'A' => 'a',
        ];
    }

    protected static function values(): array
    {
        return [
            'B' => 2
        ];
    }
}
```

The benefit of making these methods instead of consts, is that the IDE gives you autocompletion when you want to override them, and you can ensure the return type is an array.

Back to `EnumDefinition`, it has an `equals` method which is used to determine whether an enum can actually be constructed based on a value.

It is smart enough to support both method names and mapped values, as well as keeping track of case-insensitive method names.

The use of `EnumDefinition` again reduces the complexity in the `Enum` class itself. Its constructor can simply accept an input value, and it will make sure an instance with the correct value is returned.

On the topic of the name `value`, @Gummibeer and I talked about how `index` is the better term which is used in many other programming languages. However in PHP, it seems like `value` is the commonly used term, both in [`SplEnum`](https://www.php.net/manual/en/class.splenum.php) and in [`myclabs/enum`](https://github.com/myclabs/php-enum)

I believe it's best to keep the terminology is close as possible to what the PHP community is used to.

## Other changes

- `is` methods are still allowed
- `getValue()` is replaced by a readonly property `$enum->label`
- `getIndex()` is replaced by a readonly property `$enum->value`
- `isEqual` is replaced by `equals`, which now takes one or more other enums
- `equalsAny` is dropped in favour of `equals`:  `$enum->equals(MyEnum::A(), MyEnum::B(), …)`

## Open considerations

- I've dropped the interface for now, since I don't see lots of added benefit, but we can discuss adding it, are there any good reasons to have it?
- `toArray` returns a map of `$value => $label`, where both value and label are the mapped equivalents when present. Is this the behaviour we want?

Gummibeer commented:
Without getting into details and ignoring your open questions for a moment my major issue/question is if we are fine with not enforcing `value` to be an `int`? 🤔

I haven't checked the code, but right now we could definitely use every scalar and possibly even array or object - which could lead to problems in some conditions or extending code, serialization and so on. 

For me the concept of enums are a super strict one - no room for interpretation, mixed options and so on. Like you said, if that's needed use states.
So I would like to enforce as much types and as strict as possible.

Regarding the const to method replacement I agree with the auto completion! I always have to check the docs, code or other enum if it's `MAP_INDEX` or `INDEX_MAP` or anything else. So I definitely feel you.
But doing it in a method allows for theoretically dynamic maps - doing strange things with function calls or even do DB queries.
Consts will require a code change that has to be reviewed and deployed. So it's saver and more enum like constant 😅🙈 and stable.
I definitely don't want to support both, but would like to argue why I've decided for constants.

For me this is like other fundamental concept implementations - in opposite to packages like activitylog or any others the benefit for the user is in the strictness of this concept implementation. The less decisions for the user the less issues/proems for us/maintainers - except "I want to be able to do XYZ" requests.

This strictness would also allow us to type-hint the protected properties.

Regarding the interface: for me that's a safety in implementation to prevent enforcing to extend the class itself. For me it's also better to type-hint an interface instead of an abstract class. But in fact this would require to switch to `label()` and `value()` methods instead of overloaded properties. Otherwise the interface API wouldn't be complete/working. And as a dev it's also a safety for me that my changes aren't breaking if I don't have to adjust the interface.

I will review the code changes later today - right now I'm not in the office.

brendt commented:
> Without getting into details and ignoring your open questions for a moment my major issue/question is if we are fine with not enforcing value to be an int? 🤔
> 
> I haven't checked the code, but right now we could definitely use every scalar and possibly even array or object - which could lead to problems in some conditions or extending code, serialization and so on.
> 
> For me the concept of enums are a super strict one - no room for interpretation, mixed options and so on. Like you said, if that's needed use states.

I see why there's value in allowing ints, but I believe string values are a must, based on our projects. Our data sets in general are small enough that string search on databases is performant enough, and based on my own experience it's way more easy to debug and manage string values than numbers. 

Take the following example, which I actually had to deal with in a project we did some consulting on a few years back. Take the concept of a `Transport`, which has one of these states (internally this state was represented by an enum)

```php
TransportState::pending(); // value 1
TransportState::on_site(); // value 2
TransportState::done(); // value 3
```

A new use case came about, one where a new state should be added called `sorting`, which happened between `on_site` and `done`. Now, we didn't build the original code, so we went ahead and added an enum value `sorting` with a value of 4.

More changes had to be made to this flow, resulting in enum values that were out of order, even negative values 😱

If this example made one thing very clear to me is that you should never rely on the enum value itself to indicate anything else but the value. Furthermore, reading the raw data in the database, it was never clear which state a transport was actually in: you had to go look up the corresponding enum value in the code.

That's an example why I think enum values should almost be stored as strings, to prevent such shenanigans from happening.

> But doing it in a method allows for theoretically dynamic maps - doing strange things with function calls or even do DB queries.

If someone does this, they are screwing up big time, and I don't think it's the package responsibility to prevent it.

> Regarding the interface: for me that's a safety in implementation to prevent enforcing to extend the class itself

I'm not really sure I got what you're saying here, custom enum classes should always extends the base `Enum` class. I probably didn't understand it right though, so feel free to further explain :)

Gummibeer commented on .github/workflows/run-tests.yml:
```diff
@@ -10,9 +10,9 @@ jobs:
-                php: [7.4, 7.3, 7.2]
+                php: [7.4]
                 dependency-version: [prefer-lowest, prefer-stable]
-                os: [ubuntu-latest, windows-latest]
```
would like to keep windows tests, if not even add macos. But at least windows sometimes behaves different than linux and this is a super basic package it should be battle-tested - also on windows.
We can discuss this again if tests are failing on windows without any reason/way to fix.^^

brendt replied:
Fair enough!

Gummibeer replied:
https://github.com/spatie/enum/pull/56/commits/c42de09faa94f814cf2a84ef111ac7747b997292

Gummibeer commented on composer.json:
```diff
@@ -23,11 +23,12 @@
     ],
     "require": {
-        "php": "^7.2",
+        "php": "^7.4",
```
✅

Gummibeer commented on .github/workflows/run-tests.yml:
```diff
@@ -10,9 +10,9 @@ jobs:
         strategy:
             matrix:
-                php: [7.4, 7.3, 7.2]
+                php: [7.4]
```
✅

Gummibeer commented on composer.json:
```diff
@@ -23,11 +23,12 @@
     },
     "require-dev": {
-        "phpunit/phpunit": "^8.0"
+        "larapack/dd": "^1.1",
```
Still not sure about this need as dependency. But I know that you do this quite often and expect it to be available at spatie - so I won't reject it as long as it's not conflicting.

brendt replied:
If it's conflicting we'll remove it!

Gummibeer commented on src/EnumDefinition.php:
```diff
@@ -0,0 +1,31 @@
+class EnumDefinition
+{
+    /** @var string|int */
+    public $value;
```
Don't make `public` properties make this class vulnerable for runtime changes?

brendt replied:
In principle, yes. Ever since PHP 7.4 we've been using public properties more often though. readonly properties are still missing unfortunately, but I feel it's not worth the overhead of adding getters, since this class is only to be used internally by the enum itself. That was my reasoning :)

Gummibeer commented on src/EnumDefinition.php:
```diff
@@ -0,0 +1,31 @@
+
+    public function equals($input): bool
+    {
+        return $this->value === $input
```
Right now it's also possible to instantiate an enum by its `value` or after renaming the `label`.
Should we drop this?

brendt replied:
I think we should drop it, at least I can't come up with cases where this would be beneficial, it only adds complexity in the package, while I think the added "flexibility" adds more confusion than anything else. If there are valid use cases that I can't think of; I'm fine with reconsidering this.

As a side note: the goal of labels is that they allow a textual representation that is allowed to change over time, without having to worry about migrating data. Allowing enums to be constructed from their label seems like partly nullifying that goal.

Gummibeer replied:
> As a side note: the goal of labels is that they allow a textual representation that is allowed to change over time, without having to worry about migrating data. Allowing enums to be constructed from their label seems like partly nullifying that goal.

:+1:

Gummibeer commented on src/EnumDefinition.php:
```diff
@@ -0,0 +1,31 @@
+        $this->label = $label;
+    }
+
+    public function equals($input): bool
```
I would like to reference your guidelines and prefer single short `if`s over this line-breaked `or` one.

brendt replied:
Absolutely right, I've changed it!

Gummibeer replied:
https://github.com/spatie/enum/pull/56/commits/991120832d9b6a11b4170d2cf158a1873792f63b

Gummibeer commented on src/EnumDefinition.php:
```diff
@@ -0,0 +1,31 @@
+
+    public string $label;
+
+    private string $methodName;
```
By having this property `private` this class will be unextendable because the key information isn't accessible for extending classes. So I would be fine with making the whole class also `final` - indicating more prominent that there's no good way to extend it and that it's only an internal value object.

brendt replied:
I totally agree, but @freekmurze doesn't, which means I'm stuck 😭

Gummibeer replied:
I know that @freekmurze doesn't like `final` which is okay for me and I also use it very very very limited in packages. But if everything is private the class is like `final` because even if you extend you start with a blank canvas - unable to call/retrieve anything from the parent.

So I'm okay with make everything `protected` to keep it extendable - for whatever reason. Or make it `final`, which I prefer, and also add `@internal` annotation to indicate that this class is like invisible and should never be used by the user.

Gummibeer commented on src/Exceptions/DuplicatedValueException.php:
```diff
@@ -1,17 +0,0 @@
-
-use InvalidArgumentException;
-
-class DuplicatedValueException extends InvalidArgumentException
```
I think that we still need this Exception and should also throw it because we still allow custom maps which aren't enforced to be unique.
But rename it to `DuplicateLabelException`.

brendt replied:
I agree!

Gummibeer commented on src/Exceptions/DuplicatedIndexException.php:
```diff
@@ -1,17 +0,0 @@
-
-use InvalidArgumentException;
-
-class DuplicatedIndexException extends InvalidArgumentException
```
I think that we still need this Exception and should also throw it because we still allow custom maps which aren't enforced to be unique.
But rename it to `DuplicateValueException`.

brendt replied:
I agree!

Gummibeer commented on src/Exceptions/InvalidIndexException.php:
```diff
@@ -1,19 +0,0 @@
-
-use InvalidArgumentException;
-
-class InvalidIndexException extends InvalidArgumentException
```
I think that we still need this exception but extend it to also allow `string` for `value`.

Gummibeer commented on src/Exceptions/InvalidValueException.php:
```diff
@@ -1,19 +0,0 @@
-
-use InvalidArgumentException;
-
-class InvalidValueException extends InvalidArgumentException
```
I would prefer this explicit exception over a type error if a non string is written into `EnumDefinition::$label`

brendt replied:
My preference is actually to use the type system to its best extent where possible.

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
-        return $this->name;
-    }
+        if (strpos($name, 'is') === 0) {
+            $other = new static(lcfirst(str_replace('is', '', $name)));
```
The `lcfirst` isn't needed because methods can start with uppercase and we also use `strtolower` before instantiating one.

brendt replied:
you're right!

Gummibeer replied:
https://github.com/spatie/enum/pull/56/commits/804e224c27ae96c2d843033603ade18329a86ef9

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
     }
 
-    public function isAny(array $values): bool
+    public function equals(Enum ...$others): bool
```
✅

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
+    public static function __callStatic(string $name, array $arguments)
     {
-        return array_column(static::resolve(), 'value');
+        return new static($name);
```
Right now you are also able to do `MyEnum::isDraft('draft')` - do we want to drop this possibility? This is useful if you want to test if any given value equals a fixed enum. For example to adjust validation rules in a request.

brendt replied:
I want this package to be focussed on the enum objects as much as possible, so that there's as little area as possible where you need to think about whether you're dealing with the object or the value — this was my main gripe with `myclabs/enum`, which is why I wanted to build this one. The more places support both the value- and object representation, the more place there are where you do need to worry about the distinction between them. 

Since there's a middleware in the laravel package which automatically transforms enums from requests to objects, I don't think we need to support the value representation. 

Looking forward to hearing your thoughts!

Gummibeer replied:
Fien for me - a lot of my questions are primary to bring up the breaking points in this release and check if they are intentional or because of the current complexity simply forgotten.
I agree that the more we only use/allow an instance of the enum the better for us and the user regarding typing.

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
     }
 
-    public function __toString(): string
+    public function __get($name)
```
I would still prefer `public function label(): string` over this overloaded read-only property. 🤔 
For me functions are clearer that they are read-only.

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
-            if (! static::isValidIndex($value)) {
-                throw new InvalidIndexException($value, static::class);
+        foreach (static::resolveDefinition() as $definition) {
+            if ($definition->equals($input)) {
```
Do we really need this definition class/object? right now it's only used as data storage after resolving until instantiating an enum.
The `EnumDefinition::equals()` method is only used once - here. I'm not sure how much we really win by introducing it instead of simply having an array in the cache array. 🤔

brendt replied:
It's a simple value object, that's right. But I think it does serve a purpose, it simplifies the `Enum` class just a little bit, _and_ allows the developer to actually know what kind of properties are available, instead of having to remember array keys. I dedicated a chapter in Laravel beyond CRUD on this topic, if you'd like to know more about my point of view: https://stitcher.io/blog/laravel-beyond-crud-02-working-with-data

Gummibeer replied:
I know this and also your point of view. 😉 
I think that we could have a good compromise if the `EnumDefinition` implements `ArrayAccess`!?
This would allow to still use some pretty helpful `array_xyz()` functions on the cache - see https://github.com/spatie/enum/pull/56#discussion_r431786607

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
+     * @return \Spatie\Enum\EnumDefinition[]|null
      */
-    public static function getAll(): array
+    private static function resolveDefinition(): array
```
✅ So happy with slimming this method(s) down! 🎉

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
-        if (static::isValidName($name)) {
-            return static::make($name);
+        foreach (static::resolveDefinition() as $definition) {
+            $array[$definition->value] = $definition->label;
```
Right now the `toArray()` method does the inverse.
```php
[
  'value' => 'index',
]
```
After renaming this would be `label` as key and `value` as value. But I agree that this could be not this usable if the label is used, like it suggests, for more readable strings.

So open for discussion here and I understand why key and value are flipped.

brendt replied:
It of course depends on when you'd want to user `toArray`, my assumption is to provide select fields in forms to select an enum value. In that case `$key => $value` is the most logical approach, since that's the way a select box would need it.

Gummibeer replied:
I think that I would use `array_keys($enum->toArray())` for a select input and run the array values through `trans()` (Laravel) or anything.
But because that would be still possible everything's fine for me. I'm just thinking about methods that only return values, labels - like the `getValues()`, `getIndices()` methods before. But these would "conflict" with the current `values()` and `labels()` mapping methods.

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
+        return [];
     }
 
-    public function jsonSerialize(): string
```
I think that allow to JSON serialize an enum is a pretty useful feature and we could still return `value` which would be the `int|string` stored into DB or used in requests as well.

brendt replied:
You're right, there's no harm in adding it.

Gummibeer replied:
https://github.com/spatie/enum/pull/56/commits/b3fb70680da2a517149e9a328e32b0289faf1d39

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,141 @@
-     *
-     * @return static
-     */
-    public static function make($value): Enumerable
```
The `make()` method should be kept as a static creation method.

brendt replied:
I'm ok with that, I'll add it

Gummibeer replied:
https://github.com/spatie/enum/pull/56/commits/e8641bc48b4e4de05e06b4f3c9e45b061922a4aa

Gummibeer commented:
https://github.com/spatie/enum/pull/56#issuecomment-631410062 agree with everything.

Gummibeer commented on src/Enum.php:
```diff
@@ -127,8 +129,16 @@ private static function resolveDefinition(): array
 
         $valueMap = static::values();
 
+        if (self::arrayHasDuplicates($valueMap)) {
```
This doesn't prevent this case:

```php
/** 
 * @method static self value_1()
 * @method static self value_2()
 */

$valueMap = [
  'value_1' => 'value_2',
]
```

Which will result in a final enum definition array of two times the `value_2` value.

brendt replied:
Do we really need to prevent that though? I can't imagine anyone every doing that.

Gummibeer replied:
My opinion: do it right/full or nothing.^^
So we can drop this check in it's whole if we only prevent one single source of failure. Having this exception and check but not really protecting from duplicates gives a false safety for me.

I think that this check would be easier with a cache array instead of the `EnumDefinition` class - or let the `EnumDefinition` implement `ArrayAccess` interface to still be able to use `array_column()` for such checks.

Gummibeer replied:
https://github.com/spatie/enum/pull/59/commits/d90cd1199c5658d23e8f898f510e6303e3a9735d

Gummibeer commented on src/Enum.php:
```diff
@@ -2,425 +2,170 @@
-        throw new BadMethodCallException('Call to undefined method '.static::class.'::'.$name.'()');
-    }
-
-    public function __toString(): string
```
I think that we should keep this method and return `value` - so the new renamed `value` - the one that could be the methodName, index or a custom string. So the identifier `value`.^^

This would allow to simply do `{{ $enum }}` or `trans("enums.status.{$enum}")` (Laravel).

Gummibeer replied:
https://github.com/spatie/enum/pull/59/commits/b599251bfee7044c802c33f2fed73f627212d628

SagarNaliyapara commented:
@brendt when this is going to release/merge?

brendt commented:
We're testing it in one of our projects right now, probably another week or two

SagarNaliyapara commented:
Thanks @brendt

Gummibeer commented:
@brendt let me know if you are done/need a review/feedback. I will happily check what you've done. :)

brendt commented:
Will do!

Gummibeer commented:
@brendt have merged the current `master` and accepted your `Enum` in it's whole to resolve conflicts.
So in code nothing has changed only conflict solved and readme, changelog & funding updated.

Gummibeer commented:
@dionysiosarvanitis in the current v3 code the static `is` methods are dropped. Could you explain a bit in more detail how/why you use them and if so why it would be good to keep them!?
If you are fine with dropping them would also be fine to get this feedback by an user who uses the dropped methods right now.

dionysiosarvanitis commented:
Hi! I use static `is` methods  because ther try to match both enum's name/value when comparing. Also, the syntax below is less noisy in my eyes. 

```php
Signer:::GENERALDIRECTOR()->getValue() === 'Director General' vs Signer::isGeneralDirector('Director General');
```

Is there any alternative if methods will be dropped?

Gummibeer commented:
@dionysiosarvanitis right now there's no comparison method like before that accepts string/int values and catches also possible exceptions to only return false.
So you are right that comparisons like this will need more userland code in the future:

```php
// v2
StatusEnum::make('draft')->isPublished();
StatusEnum::isPublished('draft');
StatusEnum::isDraft('hamburg');

// v3
try {
    return StatusEnum::make('draft')->equals(StatusEnum::make('hamburg'));
} catch(Exception $ex) {
    return false;
}
```

@brendt I beliefe that simplifying the package is a good idea! But primary the changes to the comparison logic make the package simpler but thee userland more complicated.

**RFC:**
* allow to pass string/int values instead of only `Enum` instances to `equals()` or add a similar method allowing this
* catch invalid value exceptions in this new method and return `false` in this case
* add support for the static `is` methods

After these changes the new *v3*  API would be similar to this:

```php
StatusEnum::make('draft')->equals('hamburg');
StatusEnum::draft()->isHamburg();
```

brendt commented:
In my opinion, you should never write `StatusEnum::make('draft')` manually; the goal of this package is to get the "raw values" as far away from the user as possible, and have them only work with the enum objects. 

That's why the laravel package will support model- and request input casts, and allow you to write this:

```php
$model->status->equals($request->input('status'));
```

You never, ever, should work with the raw values directly. If you want to it still, I'm think that spatie/enum v3 simply isn't the right solution for your problem.

I just discussed this as well with my colleague Ruben who has been using this version in one of our largest projects, and this really has never been an issue for us: we cast enum values to enum objects at the entry points of our application (model casts, request layer or by using DTOs), and never worry about the raw value anymore afterwards. It's a design decision that really fits our workflow, and we're not going to change it. 

I you want to read more about my point of view on the topic, I'd suggest taking a look at this post I wrote about it: https://stitcher.io/blog/laravel-beyond-crud-02-working-with-data

In summary: I realise some people would like more flexibility, but my goal with the package is to remove as much magic and flexibility as possible, in favour of strictness, allowing us to really use the type system and static analysis tot its full extent.

Gummibeer commented:
What should break by enforcing `__toString()` to return a string?
https://github.com/spatie/enum/pull/56/commits/1877afbbefb4616fda0be97bc6ca778b86aa9e29

Only the PHP8 `Stringable` interface would change this behavior. Otherwise it's well documented and expected that `__toString()` returns, well, a string.
https://www.php.net/manual/en/language.oop5.magic.php#object.tostring
```php
public __toString ( void ) : string
```
> This method must return a string, as otherwise a fatal `E_RECOVERABLE_ERROR` level error is emitted.

Gummibeer commented:
Why have you removed the native PHP return type hint in favor of a doc-block one?
https://github.com/spatie/enum/pull/56/commits/b96bb510e753c2921beaa3b99e1e7a2cf028f283#diff-a6876695a1a28a81719fa474346113bfL38
